### ISSUE TITLE
Dict keys and set always compare to false

### ISSUE THREAD

**Issue by NeilGirdhar** [microsoft/pyright#4695]
```python
x = {'a': 1, 'b': 2}
c = {'a', 'b'}
print(x.keys() == c)  # error: Condition will always evaluate to False since the types "dict_keys[str, int]" and "set[str]" have no overlap (reportUnnecessaryComparison)
```
However, both `KeysView` and `set` inherit from `AbstractSet` in the typeshed, which defines the `__eq__` that they use with each other.

**Comment by NeilGirdhar** [comment]
(Both `KeysView` and `set` are `AbstractSet` in the typeshed.)

**Comment by erictraut** [comment]
The current behavior is as designed. The `reportUnnecessaryComparison` check is known to potentially produce false positive errors because it doesn't have access to the internal implementation of `__eq__` methods. It assumes that if two classes have no overlap (i.e. neither is assignable to the other), then they are not comparable. This is generally a good heuristic, but it can fail, as in this case.

This is a strict-mode check. You can choose to disable it globally for your project or on a per-file basis if you find that it's too noisy. You can also use `# type: ignore` or `# pyright: ignore`. Another option (the one that I would use myself) is to modify the code so it works with the check.

**Comment by NeilGirdhar** [comment]
Okay thanks for the detailed explanation!
